airbyte-integrations/connectors/source-instagram/source_instagram/client/api.py

```diff
@@ -0,0 +1,268 @@
+        yield self._extend_record(self._api.account, fields=fields)
+
+    def _extend_record(self, ig_user, fields) -> Dict:
+        return ig_user.api_get(fields=fields).export_all_data()
+
+
+class UserLifetimeInsightsAPI(StreamAPI):
+    LIFETIME_METRICS = ["audience_city", "audience_country", "audience_gender_age", "audience_locale"]
+    period = "lifetime"
+
+    buffer_days = 29
+    days_increment = 1
+
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        account_id = self._api.account.get("id")
+        for insight in self._get_insight_records(params=self._params()):
+            yield {
+                "id": account_id,
+                "metric": insight.get("name"),
+                "date": insight.get("values")[0]["end_time"],
+                "value": insight.get("values")[0]["value"],
+            }
+
+    def _params(self) -> Dict:
+        return {"metric": self.LIFETIME_METRICS, "period": self.period}
+
+    def _get_insight_records(self, params) -> Iterator[Any]:
+        return self._api.account.get_insights(params=params)
+
+
+class UserInsightsAPI(IncrementalStreamAPI):
+    METRICS_BY_PERIOD = {
+        "day": [
+            "email_contacts",
+            "follower_count",
+            "get_directions_clicks",
+            "impressions",
+            "phone_call_clicks",
+            "profile_views",
+            "reach",
+            "text_message_clicks",
+            "website_clicks",
+        ],
+        "week": ["impressions", "reach"],
+        "days_28": ["impressions", "reach"],
+        "lifetime": ["online_followers"],
+    }
+
+    state_pk = "date"
+
+    buffer_days = 29
+    days_increment = 1
+
+    def __init__(self, api):
+        super().__init__(api=api)
+        self._state = api._start_date
+
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        account_id = self._api.account.get("id")
+        for params_per_day in self._params():
+            insight_list = []
+            for params in params_per_day:
+                insight_list += self._get_insight_records(params=params)
+            if not insight_list:
+                continue
+
+            insight_record = {"id": account_id}
+            for insight in insight_list:
+                key = (
+                    f"{insight.get('name')}_{insight.get('period')}"
+                    if insight.get("period") in ["week", "days_28"]
+                    else insight.get("name")
+                )
+                insight_record[key] = insight.get("values")[0]["value"]
+                if not insight_record.get("date"):
+                    insight_record["date"] = insight.get("values")[0]["end_time"]
+            yield self.state_filter(insight_record)
+
+    def _params(self) -> Iterator[List]:
+        buffered_start_date = max(self._state.add(minutes=1), pendulum.now().subtract(days=self.buffer_days))
+        end_date = pendulum.now()
+
+        while buffered_start_date <= end_date:
```

Comment:
can you remind me why we do 1 day increments instead of getting data for the whole period?

Reply:
On Facebook, the statistics step is always 1 day. And I did it for the convenience of the developer.
Because, otherwise, it will need to make 3 requests to get all the data, and then there will be a lot of data transformations and filtering them by date in order to eventually get objects of the type:
```
[
  {
    "date": "2021-02-18",
    "metric_1": ...,
    "metric_2": ...,
    ...
  },
  {
    "date": "2021-02-19",
    "metric_1": ...,
    "metric_2": ...,
    ...
  },
  ...
]
```
Also, some metrics may not be available for a certain date - this expands the number of checks.

I thought about how you suggest, but decided that the code looks clearer in its current form. And also we get the data for a maximum of 30 days, so it does not become a bottleneck.

(unless you think it’s better to keep it in the same variant, please say to me about this)

Reply:
Your current approach is good 👌🏼